services/gundeck/src/Gundeck/API/Public.hs

```diff
@@ -228,28 +99,45 @@ listTokensH (uid ::: _) =
 --
 -- (arianvp): I am not sure why it is convenient for clients to distinct
 -- between these two cases.
-paginateH :: JSON ::: UserId ::: Maybe ByteString ::: Maybe ClientId ::: Range 100 10000 Int32 -> Gundeck Response
-paginateH (_ ::: uid ::: sinceRaw ::: clt ::: size) = do
-  Notification.PaginateResult gap page <- Notification.paginate uid (join since) clt size
-  pure . updStatus gap . json $ (page :: Public.QueuedNotificationList)
+paginateUntilV2 ::
+  UserId ->
+  Maybe Public.RawNotificationId ->
+  Maybe ClientId ->
+  Maybe (Range 100 10000 Int32) ->
+  Gundeck Public.GetNotificationsResponse
+paginateUntilV2 uid mbSince mbClient mbSize = do
+  let size = fromMaybe (unsafeRange 1000) mbSize
+  Notification.PaginateResult gap page <- Notification.paginate uid (join since) mbClient size
+  pure $
+    if gap
+      then Public.GetNotificationsWithStatusNotFound page
+      else case since of
+        Just (Just _) -> Public.GetNotificationsSuccess page
+        Nothing -> Public.GetNotificationsSuccess page
+        Just Nothing -> Public.GetNotificationsWithStatusNotFound page
   where
-    since :: Maybe (Maybe NotificationId)
-    since = parseUUID <$> sinceRaw
-    parseUUID :: ByteString -> Maybe NotificationId
-    parseUUID = UUID.fromASCIIBytes >=> isV1UUID >=> pure . Id
+    since :: Maybe (Maybe Public.NotificationId)
+    since = parseUUID <$> mbSince
+
+    parseUUID :: Public.RawNotificationId -> Maybe Public.NotificationId
+    parseUUID = (UUID.fromASCIIBytes . Public.unRawNotificationId) >=> isV1UUID >=> pure . Id
+
     isV1UUID :: UUID -> Maybe UUID
     isV1UUID u = if UUID.version u == 1 then Just u else Nothing
-    updStatus :: Bool -> Response -> Response
-    updStatus True = setStatus status404
-    updStatus False = case since of
-      Just (Just _) -> id
-      Nothing -> id
-      Just Nothing -> setStatus status404
-
-getByIdH :: JSON ::: UserId ::: NotificationId ::: Maybe ClientId -> Gundeck Response
-getByIdH (_ ::: uid ::: nid ::: cid) =
-  json @Public.QueuedNotification <$> Notification.getById uid nid cid
 
-getLastH :: JSON ::: UserId ::: Maybe ClientId -> Gundeck Response
-getLastH (_ ::: uid ::: cid) =
-  json @Public.QueuedNotification <$> Notification.getLast uid cid
+paginate ::
+  UserId ->
+  Maybe Public.NotificationId ->
+  Maybe ClientId ->
+  Maybe (Range 100 10000 Int32) ->
+  Gundeck (Maybe Public.QueuedNotificationList)
+paginate uid mbSince mbClient mbSize = do
+  let size = fromMaybe (unsafeRange 1000) mbSize
+  Notification.PaginateResult gap page <- Notification.paginate uid mbSince mbClient size
+  pure $ if gap then Nothing else Just page
+
+getById :: UserId -> Public.NotificationId -> Maybe ClientId -> Gundeck (Maybe Public.QueuedNotification)
+getById = Notification.getById
+
+getLastNotification :: UserId -> Maybe ClientId -> Gundeck (Maybe Public.QueuedNotification)
+getLastNotification = Notification.getLast
```

Comment:
both of these functions are defined 3 times, in 3 different modules, and only the last definition isn't a synonym.  i think it's better to call the real definitions in `Data` and remove the other two.